point_cloud_client/src/lib.rs

```diff
@@ -14,19 +14,88 @@ enum PointClouds {
 pub struct PointCloudClient {
     point_clouds: PointClouds,
     aabb: Aabb3<f64>,
-    pub num_points_per_batch: usize,
-    pub num_threads: usize,
-    pub buffer_size: usize,
+    num_points_per_batch: usize,
+    num_threads: usize,
+    buffer_size: usize,
 }
 
 impl PointCloudClient {
-    pub fn new(locations: &[String], data_provider_factory: DataProviderFactory) -> Result<Self> {
-        if locations.is_empty() {
+    pub fn bounding_box(&self) -> &Aabb3<f64> {
+        &self.aabb
+    }
+
+    pub fn for_each_point_data<F>(&self, point_query: &PointQuery, mut func: F) -> Result<()>
+    where
+        F: FnMut(PointsBatch) -> Result<()>,
+    {
+        match &self.point_clouds {
+            PointClouds::Octrees(octrees) => {
+                let mut parallel_iterator = ParallelIterator::new(
+                    octrees,
+                    point_query,
+                    self.num_points_per_batch,
+                    self.num_threads,
+                    self.buffer_size,
+                );
+                parallel_iterator.try_for_each_batch(&mut func)?;
+            }
+            PointClouds::S2Cells(s2_cells) => {
+                let mut parallel_iterator = ParallelIterator::new(
+                    s2_cells,
+                    point_query,
+                    self.num_points_per_batch,
+                    self.num_threads,
+                    self.buffer_size,
+                );
+                parallel_iterator.try_for_each_batch(&mut func)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct PointCloudClientBuilder<'a> {
+    locations: &'a [String],
+    data_provider_factory: DataProviderFactory,
+    num_points_per_batch: usize,
+    num_threads: usize,
+    buffer_size: usize,
+}
+
+impl<'a> PointCloudClientBuilder<'a> {
+    pub fn new(locations: &'a [String], data_provider_factory: DataProviderFactory) -> Self {
+        Self {
+            locations,
+            data_provider_factory,
+            num_points_per_batch: NUM_POINTS_PER_BATCH,
+            num_threads: num_cpus::get() - 1,
```

Comment:
silly question: what would happen if `num_cpus::get()` returns `1` (I don't know if it's even possible in XXI century :P) and, consequently, `num_thread = 0`? Wouldn't something stop working down the road?

Reply:
Good catch, I changed it everywhere to `std::cmp::max(1, num_cpus::get() - 1)`